src/db.c

```diff
@@ -1354,25 +1373,83 @@ void scanCommand(client *c) {
 
 void dbsizeCommand(client *c) {
     redisDb *db = c->db;
-    unsigned long long int size = dbSize(db, MAIN_DICT);
+    unsigned long long int size = dbSize(db, DICT_MAIN);
     addReplyLongLong(c, size);
 }
 
-unsigned long long int dbSize(redisDb *db, unsigned dictType) {
-    return db->db_type[dictType].key_count;
+unsigned long long int dbSize(redisDb *db, dbKeyType keyType) {
+    return db->sub_dict[keyType].key_count;
 }
 
-unsigned long dbSlots(redisDb *db, unsigned dictType) {
+unsigned long dbSlots(redisDb *db, dbKeyType keyType) {
     unsigned long slots = 0;
     dict *d;
-    dbIterator dbit;
-    dbIteratorInit(&dbit, db, dictType);
-    while ((d = dbIteratorNextDict(&dbit))) {
+    dbIterator *dbit = dbIteratorInit(db, keyType);
+    while ((d = dbIteratorNextDict(dbit))) {
         slots += dictSlots(d);
     }
+    zfree(dbit);
     return slots;
 }
 
+size_t dbMemUsage(redisDb *db, dbKeyType keyType) {
+    size_t mem = 0;
+    unsigned long long keys_count = dbSize(db, keyType);
+    mem += keys_count * dictEntryMemUsage() +
+            dbSlots(db, keyType) * sizeof(dictEntry*) +
+            db->dict_count * sizeof(dict);
+    if (keyType == DICT_MAIN) {
+        mem+=keys_count * sizeof(robj);
+    }
+    return mem;
+}
+
+dictEntry *dbFind(redisDb *db, void *key, dbKeyType keyType){
+    int slot = getKeySlot(key);
+    if (keyType == DICT_MAIN)
+        return dictFind(db->dict[slot], key);
+    else if (keyType == DICT_EXPIRES)
+        return dictFind(db->expires[slot], key);
+    else
+        serverAssert(0);
+}
+
+/* 
+ * This method is used to iterate over the elements of the entire database specifically across slots.
+ * It's a three pronged approach. 
+ *
+ * 1. It uses the provided cursor `v` to retrieve the slot from it. 
+ * 2. If the dictionary is in a valid state checked through the provided callback `dictScanValidFunction`,
+ *    it performs a dictScan over the appropriate `keyType` dictionary of `db`.
+ * 3. If the slot is entirely scanned i.e. the cursor has reached 0, the next non empty slot is discovered. 
+ *    The slot information is embedded into the cursor and returned.
+ */
+unsigned long long dbScan(redisDb *db, dbKeyType keyType, unsigned long long v, dictScanFunction *fn, dictScanValidFunction *valid, void *privdata) {
+    dict *d;
+    unsigned long long cursor = 0;
+    /* During main dictionary traversal in cluster mode, 48 lower bits in the cursor are used for positioning in the HT.
+     * Following 14 bits are used for the slot number, ranging from 0 to 2^14-1.
+     * Slot is always 0 at the start of iteration and can be incremented only in cluster mode. */
+    int slot = getAndClearSlotIdFromCursor(&v);
+    if (keyType == DICT_MAIN)
+        d = db->dict[slot];
+    else if (keyType == DICT_EXPIRES)
+        d = db->expires[slot];
+    else
+        serverAssert(0);
+    if (valid == NULL || valid(d) == C_OK) {
```

Comment:
shouldn't we move to the next slot when the `valid()` return ERR?
otherwise, the loop in expire.c will just keep looping, instead of maybe breaking out?
am i missing something?
p.s. i suppose we should make sure this is covered by tests.

Reply:
Damn I missed this case. Let me handle it.

Reply:
It's slightly difficult to pause the scanning for a slot while it is in progress and undergoes a large data cleanup to have 99% buckets empty. 

I've added a test case to verify the skip logic is handled and the slot iteration continues.